isaac2/Assets/Scripts/Health.cs

```diff
@@ -0,0 +1,176 @@
+    {
+        if (health == 0) // ALREADY DEAD
+        {
+            return health;  // prevent healing of dead entities
+        }
+        else if (newHealth > maxHealth) // OVERHEAL
+        {
+            health = maxHealth;
+            // overheal - could block use of heal item or convert heal to something else
+        }
+        else if (newHealth <= 0) // YOU DIED
+        {
+            health = 0;
+            // TODO: initiate death stuff
+            
+            // obviously death differs between players and enemies so might need to
+            // have separate health scripts for each to account for that?
+            // enemies - either despawn or just chill as a corpse
+            // player - uhhhhhhhhhhhhhhhh
+        }
+        else
+        {
+            health = newHealth;
+        }
+        return health;
+    }
+
+    /// <summary>
+    /// Increment current health by given + or - delta amount.
+    /// Calls setHealth which handles death, overheal, etc.
+    /// Delta is 1 by default (enemy heals by 1 HP).
+    /// Returns new health in case caller needs (detect death, etc.).
+    /// </summary>
+    /// <param name="delta"></param>
+    /// <returns></returns>
+    public int IncrementHealth(int delta = 1)
+    {
+        return SetHealth(health + delta);
+    }
+
+    /// <summary>
+    /// Decrement current health (damage) by given delta amount.
+    /// Delta is 1 by default (entity takes 1 HP damage).
+    /// Returns new health in case caller needs (detect death, etc.).
+    /// </summary>
+    /// <param name="delta"></param>
+    /// <returns></returns>
+    public int Damage(int delta = 1)
+    {
+        return IncrementHealth(-delta); 
+        // convention question: is damage param negative or positive?
+        // TODO: visual cues?
+    }
+
+    /// <summary>
+    /// Increment current health (heal) by given delta amount.
+    /// Delta is 1 by default (entity heals 1 HP).
+    /// Returns new health in case caller needs.
+    /// Might want to return overheal flag???
+    /// </summary>
+    /// <param name="delta"></param>
+    /// <returns></returns>
+    public int Heal(int delta = 1)
+    {
+        return IncrementHealth(delta);
+        // TODO: visual cues?
+    }
+
+    /// <summary>
+    /// Set current health to the max health / fully heal entity.
+    /// Returns current maxHealth (probably unnecessary).
+    /// </summary>
+    /// <returns></returns>
+    public int ResetHealth()
+    {
+        return SetHealth(maxHealth);
+        // TODO: visual cues?
+    }
+
+    /// <summary>
+    /// Set max health to given amount.
+    /// Could be used for powerups etc.
+    /// Returns maxHealth in case caller needs.
+    /// </summary>
+    /// <param name="newMaxHealth"></param>
+    /// <returns></returns>
+    public int SetMaxHealth(int newMaxHealth) 
+    {
+        if (newMaxHealth <= 0)  // THATS ILLEGAL
+        {
+            // ERROR!! resist change (or maybe set to 1?)
```

Comment:
I think we should truncate to 0.

Reply:
I think this is done? Unless there is a better way to do this.